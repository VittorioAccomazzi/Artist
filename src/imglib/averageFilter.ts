import Histogram from './histogram'

export default class AverageFilter {

    /**
     * filter in place the histogram with a moving average.
     * @param inHist input histogram
     * @param kSize filter size
     */
    static Run(inHist : Histogram, kSize : number)  {
        let fBins = inHist.histogramBins
        if( kSize > 0 ){
            let inBins = inHist.histogramBins
            let ouBins : number [] = Array(inBins.length).fill(0)
            ouBins.filter((v,index)=>{
                let sum = 0
                let count= 0
                for(let i=index-kSize; i<= index+kSize; i++){
                    if( i>=0 && i<inBins.length){
                        count++
                        sum += inBins[i]
                    }
                }
                ouBins[index]=Math.floor(sum/count)
            })
            fBins = ouBins
        }

        return new Histogram(fBins)
    }
}