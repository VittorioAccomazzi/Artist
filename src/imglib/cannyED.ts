import {ImageFactory, Image2D, ImagePixels, ImageUint8} from './imagebase'
import Gradient from './gradient'

export default class CannyEdgeDetection {

    /*
    Detect(inImage:Image2D, lowThr : number, highThr : number ) : ImageUint8 {
        let nsImage = NonMaxSuppressedImage(inImage)
    }


    private NonMaxSuppressedImage ( inImage : Image2D ) : ImageUint16 {

    }
    */

}