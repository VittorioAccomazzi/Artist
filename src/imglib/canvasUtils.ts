import {ImageFactory, ImageUint8} from './imagebase'

export default class CanvasUtils {

    /**
     *  Generate the canvas's channels (red, green, blue and alpha) in separate images.
     * @param canvas 
     */
    static Split( canvas : HTMLCanvasElement ) : [ ImageUint8, ImageUint8, ImageUint8, ImageUint8 ] {
        let width = canvas.width
        let height= canvas.height
        let ctx = canvas.getContext('2d') as CanvasRenderingContext2D
        let data= ctx.getImageData(0,0,width,height)
        let imgR= ImageFactory.Uint8(width,height)
        let imgG= ImageFactory.Uint8(width,height)
        let imgB= ImageFactory.Uint8(width,height)
        let imgA= ImageFactory.Uint8(width,height)
        let nPixels = width*height
        let i =0
        let cPixels = data.data
        let rPixels = imgR.imagePixels
        let gPixels = imgG.imagePixels
        let bPixels = imgB.imagePixels
        let aPixels = imgA.imagePixels
    
        for( let p=0; p<nPixels; p++){
            rPixels[p]=cPixels[i++]
            gPixels[p]=cPixels[i++]
            bPixels[p]=cPixels[i++]
            aPixels[p]=cPixels[i++]
        }
    
        return [ imgR, imgG, imgB, imgA ]
    }

    /**
     * generate a canvas object using the images provided as channels.
     * @param imgR 
     * @param imgG 
     * @param imgB 
     * @param imgA 
     */
    static Compose( imgR : ImageUint8, imgG : ImageUint8, imgB : ImageUint8, imgA? : ImageUint8 ): HTMLCanvasElement {
        let width = imgR.width
        let height= imgR.height
        let canvas = document.createElement('canvas')
        canvas.width = width
        canvas.height= height
        let ctx = canvas.getContext('2d') as CanvasRenderingContext2D
        let data= ctx.getImageData(0,0,width,height)
        let nPixels = width*height
        let i =0
        let cPixels = data.data
        let rPixels = imgR.imagePixels
        let gPixels = imgG.imagePixels
        let bPixels = imgB.imagePixels
        let aPixels = imgA?.imagePixels
    
        for( let p=0; p<nPixels; p++){
            cPixels[i++] = rPixels[p]
            cPixels[i++] = gPixels[p]
            cPixels[i++] = bPixels[p]
            cPixels[i++] = aPixels ? aPixels[p] : 255
        }

        ctx.putImageData(data,0,0)

        return canvas
    }

}