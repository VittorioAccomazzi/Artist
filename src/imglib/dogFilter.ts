const assert = require('assert')
import GaussianFilter from './gaussianFilter'
import {Image2D, ImageFactory, ImageFloat32} from './imagebase'

export default class DifferenceOfGaussian {


    /**
     * Applies the difference of gaussian, where sSigma < lSigma.
     * See Real-time video abstraction 
     *   https://www.researchgate.net/publication/220184181_Real-time_video_abstraction
     * section 3.2 on page 4
     * @param inImage input image
     * @param sSigma small sigma. It determines the spatial scale for edge detection
     * @param lSigma large sigma.
     * @param sensitivity sensitivity of the edge detector. For small values, less noise is detected, but real edges become less prominent. As it approaches 1, the filter becomes increasingly unstable. Typical value 0.98
     * @param sharpness determines the sharpness of edge representations, typically in [0.75,5.0]
     */
    static Run( inImage : Image2D, sSigma : number, lSigma : number, sensitivity : number = 0.98, sharpness : number  = 3 ) :  ImageFloat32 {
        let width = inImage.width
        let height= inImage.height
        let ouImage = ImageFactory.Float32(width, height)

        let sImage = inImage.convertTo(inImage.imageType) // copy
        let lImage = inImage.convertTo(inImage.imageType) // copy

        GaussianFilter.Run(sImage,sSigma)
        GaussianFilter.Run(lImage,lSigma)

        let sPixels = sImage.imagePixels
        let lPixels = lImage.imagePixels
        let oPixels = ouImage.imagePixels

        sPixels.forEach((val : number , index : number )=>{
            let v = val - sensitivity * lPixels[index]
            let r = 1
            if( v <0 ) r = 1+ Math.tanh(sharpness*v)
            oPixels[index]=r
        })
        
        return ouImage
    }
}
