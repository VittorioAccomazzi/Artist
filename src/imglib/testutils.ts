import { Canvas, createCanvas, loadImage } from 'canvas';
import {Image2D, isImage} from './imagebase'
import Histogram from './histogram'
import CanvasUtils, {SeqCanvas} from './canvasUtils'
import {imageHash} from 'image-hash'
import * as fs from 'fs'
import * as path from 'path'
import { createTextChangeRange } from 'typescript';


export const tmpFolder = 'tmp'
export const testImage1 = 'src/imglib/testimages/test1.jpg'
export const testImage2 = 'src/imglib/testimages/test2.jpg'
export const testImages = [ testImage1, testImage2]

export async function hash( image : Canvas | Image2D ) : Promise<string> {
    let canvas =  isImage(image) ? toCanvas(image) : image
    let b = canvas.toBuffer('image/png');
    let name = `testImage-${canvas.width}x${canvas.height}.png`
    return new Promise((res, rej)=>{
        imageHash({
            ext: 'image/png',
            data: b,
            name: name
        }, 128, true, (err: Error, val : string )=>{
            if( err ) rej(err)
            res(val)
        })
    })
}

export async function* getCanvases() : AsyncGenerator<[Canvas, CanvasRenderingContext2D, string]>{
    for( let img of testImages) {
        let image = await loadImage(img)
        let canvas= createCanvas(image.width, image.height);
        let ctx = canvas.getContext('2d')
        ctx.drawImage(image, 0, 0)
        yield [canvas, ctx, path.basename(img,'.jpg')]
    }
}

export function toCanvas(image : Image2D) : Canvas {
    let max = image.maxValue()
    let min = image.minValue()
    let img = image

    if ( max > 255 || min <0 ){
        // no need to convert
        let range = max - min
        let scale = range > 0 ? 254/range : 1
        scale = Math.min(1, scale) // to not scale up.
        let offset = -scale * min
        img = image.convertTo('Uint8', scale, offset)
    }  
    let width = img.width
    let height= img.height
    let canvas= createCanvas(width, height)
    let ctx = canvas.getContext('2d')
    let data = ctx.getImageData(0,0, width, height)
    let pixels= img.imagePixels
    let ptr =0
    for(let y=0; y<height; y++){
        let yOffset = y*width
        for( let x=0; x<width; x++){
            let p = Math.floor(pixels[yOffset+x])
            data.data[ptr++] = p // Red
            data.data[ptr++] = p // Green
            data.data[ptr++] = p // blue
            data.data[ptr++] = 255 // alpha
        }
    }
    ctx.putImageData(data,0,0)
    return canvas
}

export async function dumpImage( image : Image2D, name? : string ){
    let nCanvas = toCanvas(image)
    dumpCanvas(nCanvas,name)
}

export async function dumpCanvas(canvas : Canvas, name? : string ) : Promise<void> {
    let fullname = path.resolve(tmpFolder, name ? name: (await hash(canvas)).substr(0,25))
    if( ! fs.existsSync(tmpFolder)){
        fs.mkdirSync(tmpFolder)
    }
    await saveCanvas(canvas, fullname)
}

export function toSeqCanvas( nCanvas : Canvas ) : SeqCanvas {
    // this is inefficient, but allows to test more code
    let hCanvas = toHTMLCanvas(nCanvas)
    return CanvasUtils.toSeq(hCanvas)
}

export function toHTMLCanvas(nCanvas : Canvas) : HTMLCanvasElement {
    let hCanvas = document.createElement('canvas')
    hCanvas.width = nCanvas.width
    hCanvas.height= nCanvas.height
    let hCtx = hCanvas.getContext('2d')
    let nCtx = nCanvas.getContext('2d')
    let nDat = nCtx.getImageData(0,0,nCanvas.width,nCanvas.height)
    hCtx?.putImageData(nDat, 0, 0)
    return hCanvas
}

export function toNodeCanvas(inCanvas : HTMLCanvasElement | SeqCanvas ) : Canvas {
    let nCanvas = null
    if( isSeqCanvas (inCanvas)  ) {
        // this is more inefficient, but allows to test more code
        let hCanvas =  document.createElement('canvas')
        CanvasUtils.fromSeq(inCanvas,hCanvas)
        nCanvas = toNodeCanvas(hCanvas)
    } else {
        nCanvas = createCanvas(inCanvas.width,inCanvas.height)
        let nCtx = nCanvas.getContext('2d')
        let hCtx = inCanvas.getContext('2d') as CanvasRenderingContext2D
        let hDat = hCtx.getImageData(0,0,nCanvas.width,nCanvas.height) 
        nCtx?.putImageData(hDat, 0, 0) 
    }
    return nCanvas
}

function isSeqCanvas(obj : any ) : obj is SeqCanvas {
    return obj.data != null && obj.width != null && obj.height != null && obj.data.length > 0 
}

async function saveCanvas( canvas : Canvas, filename : string ) : Promise<void>{
    return new Promise( (res,rej)=>{
        const out = fs.createWriteStream(`${filename}.png`)
        const stream = canvas.createPNGStream()
        stream.pipe(out)
        out.on('finish', () =>  res())
    })
}

const maxHeight = 512
export  function displayHisto(histo : Histogram , cMap? : string []) : Canvas {
    let maxVal= histo.maxValue
    let bins  = histo.histogramBins
    let scale = Math.min( 1, maxHeight/maxVal)
    let width = bins.length
    let height= Math.floor(scale*maxVal)
    let canvas = createCanvas(width,height)
    let ctx = canvas.getContext('2d')
    ctx.fillStyle="#FFFFFFFF"
    ctx.fillRect(0,0,width,height)
    bins.forEach((v,i)=>{
        let col =  cMap ? cMap[i] : "#0000FFFF"
        ctx.beginPath()
        ctx.strokeStyle= col
        let scaled = Math.floor(v * scale)
        ctx.moveTo(i,height)
        ctx.lineTo(i,height-scaled)
        ctx.stroke()
    })
    return canvas
}

export default {}