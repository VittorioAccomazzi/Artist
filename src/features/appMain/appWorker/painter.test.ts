import Painter from './painter'
import { dumpCanvas, getCanvases, toSeqCanvas, hash, toNodeCanvas } from '../../../imglib/testutils'
import {SeqCanvas} from '../../../imglib/canvasUtils'

xtest('Painter quality validation', async ()=>{
    const medianKernerl = 5
    for await ( const [canvas,ctx,name] of getCanvases() ) {
        let  sCanvas : SeqCanvas | null = toSeqCanvas(canvas)
        let painter = new Painter(sCanvas)
        let num =0

        do {
            let canvas = toNodeCanvas(sCanvas)
            //await dumpCanvas(canvas,`painter ${name} ${num}`)
            num ++
            let hsh = await hash(canvas)
            expect( hsh ).toMatchSnapshot()
            sCanvas = painter.next()
        } while(sCanvas)
    }
},60*60*1000)