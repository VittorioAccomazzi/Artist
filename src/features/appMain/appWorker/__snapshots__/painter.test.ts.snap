// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Painter quality validation 1`] = `"0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000147d9000000000000000000000000005fffffff000000000000000000000004fffffffdf80000000000000000000003fffffffffd200000000000000000007fffffffffffc0000000000000000007fffffffffffff900000000000000003ffffffffffffffff0000000000000007fffffffffffffffff80000000000005ffffffffffffffffffc004000000003ffffffffffffffffffffc0800000000fffffffffffffffffffffff800000000fffffffffffffffffffffff40000007fffffffffffffffffffffffffc00000fffffffffffffffffffffffffff6600bfffffffffffffffffffffffffffd700ffffffffffffffffffffffffffffff90ffffffffffffffffffffffffffffffc0ffffffffffffffffffffffffffffffe0fffffffffffffffffffffffffffffff2ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffdfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff000000000000000000000060000000000000000000000000000003e0000000000000000000003b7b000001c000000000000000000003fffffa0001000000000000000000001ffffffe1f00000000000000000000007fffffffcc1800000000000000000003ffffffffff7e0000000000000000000ffffffffffff00000000000000000007fffffffffff80000000000000000001ffffffffffff80000000000000000001ffffffffffffe0000000000000000007fffffffffffff000000000000000001ffffffffffffffc0000000000000000fffffffffffffffe0000000000000001ffffffffffffffffc000000000000003fffffffffffffffff80000000000001fffffffffffffffffff8000000000003ffffffffffffffffffff000000000007ffffffffffffffffffff00000000001fffffffffffffffffffffe0000000003ffffffffffffffffffffff800000000ffffffffffffffffffffffff00000003ffffffffffffffffffffffffc0000007ffffffffffffffffffffffffc000001ffffffffffffffffffffffffffc0000fffffffffffffffffffffffffef80007ffffffffffffffffffff7fffd000000fffffffffffffffffffe620130000000fffffffffffffffffff2000000000000ffffffffffffffffff80000000000000fffffffffffffc000000000000000000fffffffffff8d0000000000000000000fffffffffffffbffffa0000000000000fffffffffffc0071f070000006000000fffffffffffc005e7048000000000000ffffffffffc000000000000000000000fffff000000000000000000000000000000000000000000000000000000000000000000000000003fffffff80004000000000000000003bfffffffffffffffff0000000012efffffffffffffffffffff000000003fffffffffffffffffffffff00000007ffffffffffffffffffffff7f00000007fffffffffffffffffc53ffff0000002ffffffffffffffffffffffffb101e1ffffffffffffffffffffffe3ffe7817fffffffffffffffffffffc00800183fffffffffffffffffffdf800000000fffffffffffffffffffffffd00000000fffffffffffffffffffffa0000000000ffffffffffffffffffffc20000000000fffffffffffffffffffffff380000000ffffffffffffffffffffe00000000000ffffffffe7ff97ffffffff4000000000ffffffffffdafffe03ffa000000000007fcffffffc3ff81fffdc000000000000dfffff00fff3ffff8000000000000000e7ebff7ffc7ffd7000000000000000001003ffc1ffffb6063f820000000000000000f3efff80007df000000000000000c08406780004f24000000000000000000003ff800ffffb80003b00000000000000000002fff00000020000000000000003801c07f80000018000000000000000ffffffffffffffffffffffffffffffffffffffffffffffffffbffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff9fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff4641cfffffffbfffffffffffffffff0ffffffffffff7cbffffffffffff801fffffffe07fffffffffffffffffffffffff803f00087ffffffffff3fffffffeffffffffffe7f87f3e7e1dffbffffffffdf800003ffffffeff90fffffffff7ff800004000ffff001fffffff07381e07d61fffff0c000065e9fe7ffffffdfff38007ff11f3c000003ff02001f8f3f0040c0000007dfc00010001c7fffffffe2000100100000010000000080000fff8bffffdf803ff0000000040000000ffffc000000000000000000187b8000f0000000000007800000008407ffffff000000000000180000000400011fff8034000000061fc0000000000f500000000a300004f9f82000000000000000000001ef80003e00000000000000000000000f70000000000000000000000000000008000000000000000000000000000000000078300000000000000000000000000fff8000300000000000000000000000f000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"`;

exports[`Painter quality validation 2`] = `"00000000000c01effdf0000000000000000000003c6000fffff0000000000000000000000fff000fe1c00000000000000000000007ffffffff07e000000000000000000003bffffffffff8000000000000000000055bfffffffff00000000000000000003ffbffffffe7c000000000000000010fffffffffffe3c000000000000000000fffffffffffffff80000000000000000fffffffffffffff800000000000000fffffffffffffffffe00000000000000fffffffffffdfffffe000000000000007ffffffffffffffff000000000000000ffffffffffffffffffe000000000008fffffffffffffffffff00000000000387fffffffffffffffffc00000000000303cfffffffffffffffc000000000000000033fffffffffffffffc00000000000000fffffffffffffffffc0000000000003fffffffffffffffff00000000000000fffffffffffffffffe000000000000000cffffffffffffffff7c00000000000000fffffffffffffffff800000000000007ffffffffffffffffff200000000000fffffffffffffffffffc000000000000ffffffffffffffffffe3c00000000000fffffffffffffffffffc00000000000fffffffffffffffffffd1c0000000000fffffffffffffffffffb8000000000000ffffffffffffffffffe600000000000fffffffffffffffffffff00000000003ffffffffffffffffffcff000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000008000000000000000000000001fff00e1c0c00000000000000000031fffffffff73c000000000000000003fffffffffffffca0000000000000000ffffffffffbfffcf000000000000000fffffffffffffffc0000000000000003fffffffffffffffff0000000000003ffffffffffffffffffffff000000001ffffffffffffffffffffffc000000027ffffffffffffffffffffff80000003d7fffffffffffffffffffff00000001c1fffffffffffffffffffffc0000001003fffffffffffffffffffffc00000080fffffffffffffffffffffffff00000c3ffffffffffffffffffffffffc000003fffffffffffffffffffffffff000000fffffffffffffffffffffffffc0000007fffffefffff000cfffffffff3ff00007fff3ffffff87ffffffffffffffff000ff7fffffffffffffffffffffffffc000fbfffffffffffc3fffffffffffff0000ffefffffffff7ffffffffffffcf00000fffffffffffbfffffffffffeff86e00040ffffffffffffffffffe7fff8000000efffd3fffffffffffff2600110000000fff7fff7fffffffffff0000000000000fc7ffffffffbf7a0038000000000000007fbfffffff888000000000000000000fffffffffff880000000000000000000fffffffffffffbffffe0000000000000fffffffffffd80f9f070000006000000fffffffffffc005e7068000000000000fffffffffff008000000000000000000fffffe00000000000000000000000000c00000000000000000000000000000000000000000000003fffffff70000000000000000000103ffffffffffffffffff0000000177fffffffffffffffffffffb0000000dfffffffffffffffffffffffa0000001ffffffffffffffffffbdfff1c00000007fffffffffffffffff953bffc0002006ffffffffffffffffffffdfda03bfffffffffffffffffffffffeba1d40fefffffffffffffffffffffffc00800087fffffffffffffffffffcb800000000ffffffffffffffffffffffff00000000fffffffffffffffffffffe0000000000ffffffffffffffffffffe20000000000fffffffffffffffffffffff380000000fffffffffffffffffffde00000000000ffffffffafff03fffff7fe4000000000ffffffffff98fffe03ff020000000000ff9fffffe63ff82fffd4000000000000bfffff09efebdfffc040200000000000d3ce1ffff4bffd7000100000000000002803ffc1dfff80071f800000000000000140433fffc10bf5c000000000000000c0820874000316c000000000000000000003ff8007fff18c002000000000000000000000fff00000060000000000000003001c05fc0000018000000000000000fffffffffffffffffffffff3f0400000fbfffffffffffffffffffff1c0000000ffffffffffffffffffffffe101000000fffffffffffffffffffffffd20000000ffbffffffffffffffffffffc30000000ffffffffffbffffffffffdffc4000000fffffffffe3fffbfdffffbffdc000000ffffffffffc0063fdffffffff8000000fbffffffdc61ff19fffffff8e0000000fbffffff0c6fcbc3fffffff000000000fbffffff6021f1e7cf3fffc000000000f7fc3fff01c7ffbf86eff90000000000fffffffb8001871f3fdf70400c0000007fffffff81e7e3c1e020004010000000fffffed3336ffbe7e008000000000000fff3fc003c60fffe02f8038004000000f301e000106000c0060c9fbc34000000c039c0000077000d1c05bfc0fc000000ffe8c007e047001cf001fb40000000007ff0c00000070052f000641000000002fff000000007001ff000000000000000fff000000006803ff000000000000000ffe00000000001eee000000000000000ffc00000000074800008000000000000ff80000000e077000310a00000000000f100000003fff3800000a00008000100c0000001c3dfe1d80c00000000000180c00000010008c81000000000000001c0000000000021ff7c00000000000001c0c0000002000012fc08000000000000000000600e000003f000000000000000000000410c000007680400020000200000"`;

exports[`Painter quality validation 3`] = `"fffe0000a5ffffffbd007fffffc00003fffe0000adfffcdf90003fffffc00000fffe00000bfffcdfc0003fffffe00000fffe00001dffffdfe1001fffffe00010fffe00001ffffe5fe0c01ffffff000e0fffe00000bfdfbcfa0001ffffff001c0fffe0000051bc007c0001ffffff80180fffe000003f9c003c0001ffffffc0100fffe000001bdc003c0001ffffffc0200fffe0000000c8001c0001ffffffe0000fffe00000004800184000ffffffe0000fffe000001bc9ff804000fffffff0000fffe00000298ffff0e000fffffff8000fffe00000903ffffc1000fffffff8000f9fe0000003bffffe0000fffffffc000f9fe000001f1fffff8000fffffff8000f1fe000001037ffffc000fffffff8000f1fe00000003fffffe000fffffff0000e1fe00000202ffffff800ff7fffc0000e1fe000004017fffffc00fe3fffc0001c1fe000000007fffffe00fe3fff8000381fe00000000fffffff00fe1fff0000781fe00000000fffffff80fe0fff0000701fe00000001fffffff80fe0ffe0000f01fe00000003fffffffe0fe07fc0001f01fe0000000fffffffff0fe07fc0003f01fe0000400ffbffffff0fe03f80007f01fe0000800ffbffffff8fe01f0000ff01fe0000801fc7ffffffcfe03f0000ff01fe0001801f8fffffffcfe07f0001ff01fe000180231fffffffefe0ff0003ff00fe000180443fffffffe7e3ff8007ff00fe000180007ffffffff7c7ff8007ff00fe00030000ffffffffff9fff800fff00fe00030001ffffffffffffff000fff00fe00030007ffffffffffffff001fff00fe00038007bfffffffffffff001fff00fe0003800f7fffffffffffff003fff00fe0003801efffffffffffffe007fff00fe0003c029fffffffffffffe007fff00fe0003c053fffffffffffffe00ffff00fe0003c087bffffffffffffc00ffff00fe0003f00f3ff1ffffffffcc01ffff00fe0003f0085c0cfe7fffff0c01ffff00fe0003d8103a0e007ffffe1803ffff00fe0003c0003400007ffff03003ffff00fe0003c860400001ffffc0e007ffff00fe0003d800800002ffff03c007ffff00fe0001e800400005fffc7f800fffff00fe0001f00000000ffff03f000fffff00fe0001f00000001ffff03e000fffff00fe0000f00000000ffff03c0007ffff00fe0000f00000001ffff8300007ffff00fe0000700000003ffff8300007ffff00fe0000600000007ffff830000fffff00fe000060000001fffff830001fffff00fe000060000003fffffc10005fffff00fe000060000007fffffc10005fffff00fe00002000000ffffffc1000bfffff00fe00000400000ffffffc1000bfffff00fe00001000001ffffffc10007fffff00fe00018800003fc1fffc10007ffffe00fe00010000007fb07ff008007ffffc00fe0003800000fe083fc00800fffffc00fe000fc00000fc003fc00803ffffff00fe001f800003f800ff000803ffffff00fe001e000003f00c3f080803ffffff00fe001e000003f00e3e080807ffffff00fe003c004007c67e3e780807ffffff00fe00700000079f977fc00807ffffff00ff01c0c0401f1fffffe0080fffffff00fe010001000f3ffffff0080fffffff00fe03c000401e1fffeff8080fffffff00fe0f0080001c3ffffff80c1fffffff00fe00000000781ffffff80c1ffff7ff00ff01000008383fffdff80c1fffffff00ff06410020703fffcff80c1fffffff00fe18010030601fffcff80c3fffffff00fe04010000c01ffe4ff8043fffffff00fe00010003c01ffe77f0063fffffff00ff00000003801ffe87f0067fffffff00ff00008003801fff03f0067fffffff01ff00008409001fffffe0067fffffff01ff04004032000fffffe0067fffffff01ff08002070001fffffe002ffffffff01ff10000060001fffffc002ffff7fff01ff000011c8001fffffc002ffff7fff01ff080007d00007c0008003ffff7fff01ff480407900007e1ff8003ffff7fff01ff90040f600003f1ff0003ffff7fff01ff00000fc00001f8fe0003ffff7fff00ff200827800001ff1e0003fffe7fff00ff000d03000000fffc0003fffe7fff00ff00011b000001fffc0003fffe7fff00ff0000d8000000fffc0003fffe7fff01ff0008780000003ff80003fffcfffe01ff200c7f0000001ff80003fffffff801ffc0062d800003fffe00031fffffe000ff800664c00007ffff8003c07fffc000ff2002c5800007ffffe003f800ff80c0ff400357a00003fffff803fe00ff00e0ffc1008f100003fffffc03ff84fe00f0ff800107100007fffffe03ffedfc0078ffc00007a80007ffffff07fffffc007cff800003d80003ffffff07fffffc003cff800001c80047ffffff87fffff8003cff800001880007ffffffc7fffff8003cff800001c80007ffffffc7fffff0003cff804009d80047ffffffc7fffff0001cff00000bf000c7ffffffe7fffff0001cff8000073800c7ffffffe7fffff0001cff8000038001c7ffffffe7ffffe0001cff800003c001cffffffff7ffffe0001cff80000720018fffffffffffffc0001cff80000130038ffffffffcffffc0001c7f80000000039ffffffff81fffc0001e7f80000800071ffffffff801ff80001e7f00000c00063ffffffff8007f00001e7f00000600063ffffffff8007f00001e7f800003400c7ffffffffc007e00001e7f8000008010fffffffffc007c00003e7f8000000000fffffffffc007800001e7f0000010003fffffffffe007000000e7f0002010007fffffffffe006000400f7f000101001ffffffffffe020000800f7e000100001ffffffffffe060001800f7f000080003fffffffffff071803000f7e000000007fffffffffff07f80300"`;

exports[`Painter quality validation 4`] = `"fffe00000070480d80047fffffc00003fffe00000038000000043fffffc0000afffe0000007f810500001fffffe00000fffe000049bff00f00041fffffe00000fffe000000ffd68c40001ffffff20060fffe000000dff00000001ffffff000c0fffe0000007fec0200001ffffff80180fffe000000edc00040028ffffffc0000fffe000000f5c00000020ffffffc0000fffe000000cc80000001bffffffe0000fffe00000103500004017ffffffe0000fffe000005ad9ff80003afffffff0000fffe00000fdfffff81018fffffff8000fffe0000cfcbffffc3018fffffffc000fffe000007fffffff5010fffffffc000f9fe000003fffffff8023fffffff8000f9fe00000513fffffc020fffffff8000f1fe00100e6bffffff341fffffff0000e1fe00080e1affffff801fe7fffe0000e1fe00000e213fffffe01fe7fffc0001c1fe00000d607fffffe01fe3fff80003c1fe000008a17ffffff01fe1fff0000781fe00001100fffffff80fe1fff0000f81fe00000104fffffff80fe0ffe0001f01fe00002303fffffffe1fe0ffc0003f01fe00004257efffffff0fe07fc0003f01fe0000c00bfbffffff1fe03f80007f01fe0000c00eebffffff8fe01f8000ff01fe00028803d7ffffffcfe03f0001ff01fe000181988fffffffcfe07f0001ff01fe000182321fffffffefe1ff8003ff01fe000188143fffffffefe3ff8007ff00fc000180007fffffffffc7ff8007ff01fe00010000ffffffffffdfff800fff01fe00030001ffffffffffffff000fff01fe00030002ffffffffffffff001fff01fe00038006bfffffffffffff003fff00fe000380097fffffffffffff003fff00fe0003c002effffffffffffe007fff00fe0003e005fffffffffffffe007fff00fe0003e0003ffffffffffffe00ffff01fe0003e0001ffffffffffffc00ffff01fe0003f0002ff9ffffffffcc01ffff01fe0003f0001c0dfe7fffff0c01ffff01fe0003f8001a0e007ffffe1803ffff01fe0003c000040400fffff03003ffff01fe0003c0000000017fffc0e007ffff01fe0003c000000002ffff83c007ffff01fe0001e000000005fffc6f800fffff01fe0001f00000001bfff83f000fffff01fe0001f000000007fff03e000fffff01fe0000f00000000ffff03c0017ffff00fe0000f00000003ffff0300017ffff00fe0000700000003ffff830002fffff00fe000060000000fffff810002fffff00fe000060000001fffff830007fffff00fe000060000003fffffc30007fffff00fe00006000000ffffffc10007fffff00fe000020000007fffffc1000ffffff00fe00000000000ffffffc1000ffffff00fe00001000001ffffffc1000ffffff00fe00010800003fc1fffc1001ffffff00fe00010000007f907ff01801fffffe00fe0003800000fe003fc00801fffffc00fe000f800000fc007f800803ffffff00fe001f800001f800ff000803fffffe00fe001e000001f00c3f080803fffff800fe001e000003e00e3e180807fffff300ff003c004007c43e3e780807ffffff00ff007000c0079f077f800807ffffff00ff01c0c0401f1fffffe0080fffffff00ff010001000f1ffffff0080fffffff00ff03c100401e3ffffff8080fffffff00ff070080001c3ffffff80c1fffffff00ff00010010383ffffff80c1ffff7ff00ff01000008383ffffff80c1fffffff00ff06400000303ffffff80c0fffffff00ff18010030603fffdff80c0fffffff00ff04010000c03ffecff0041fffffff00ff00010002403ffefff0061fffffff00ff00800003803ffe87e0061fffffff00ff00108003803fff87e0061fffffff00ff0000c409003fffffc0063fffffff00ff04014032003fffffc0063fffffff00ff08006070003fffffc0027fffffff00ff00002060003fffff80023fff7fff00ff000011c8003fffff0002bfff7fff00ff000607d0003fc0108003ffff7fff01ff08040790000fffff0003ffff7fff00ff90040f200007ffff0003ffff7fff00ff00040f400001fffe0003fffe7fff00ff200927800001ff1e0003fffe7fff00ff000d87000001fffc0003fffe7fff00ff00039b000001fffc0003fffc7fff00ff0000d8000000fff80003fffc7fff00ff0008f80000001ff80003fff0fffe01ff300cff0000001ff80003bffffff000ffc0063d800003fffe000303ffffe000ff80066cc00007ffff8003403fffc000ff20024d800007ffffe003f000ff80c0ff60015fa00007fffff803fc00ff00e0ffc1008f500007fffffc03ff04fe00f0ff80010f500007fffffe07ffe4fc00f8ffc00007a80007ffffff07fff7fc0078ff80200bd80007ffffff07fff7fc003cff802005c80047ffffff87fff7f8003cff802019980007ffffffc7ffeff8003cff802019c80007ffffffc7ffdff0403cff804009dc00c7ffffffc7ffdff0001cff00000bf000c7ffffffe7ffdff0001cff800007b800c7ffffffe7ff9ff0001cff8000038001c7ffffffe7fffff0001cff800003c001cffffffff7fffff0001cff80000720038ffffffff7ffffe0001cff80000130038ffffffff8ffffe0001cff80000800011ffffffff81fffc0001cff00001800071ffffffff801ff80001e7f00000c00061ffffffff8007f80001e7f000006000a3ffffffff8007f00001e7f00000340047ffffffffc007e00001e7e00000080107ffffffffc007c00003e7f0000000000fffffffffc007800001e7f0000010003fffffffffe007000001e7f0002018007fffffffffe000000400f7f0001010007fffffffffe000000400f7e000100007ffffffffffe000000800f7e00008040ffffffffffff01b801800f7e000000017fffffffffff03f80300"`;
